brain/circuit.cpp

```diff
@@ -396,36 +455,63 @@ neuron::Morphologies Circuit::loadMorphologies( const GIDSet& gids,
                                                 const Coordinates coords ) const
 {
     const URIs& uris = getMorphologyURIs( gids );
-    neuron::Morphologies result;
-    result.reserve( uris.size( ));
 
-    if( coords == COORDINATES_GLOBAL )
+    // < GID, hash >
+    Strings gidHashes;
+    gidHashes.reserve( uris.size( ));
+    std::set< std::string > hashes;
+    for( size_t i = 0; i < uris.size(); ++i )
     {
-        const Matrix4fs& transforms = getTransforms( gids );
-        for( size_t i = 0; i < uris.size(); ++i )
+        const URI& uri = uris[i];
+        std::string hash( uri.getPath( ));
+
+        if( coords == COORDINATES_GLOBAL )
         {
-            const URI& uri = uris[i];
-            const brion::Morphology raw( uri.getPath( ));
-            result.push_back( neuron::MorphologyPtr(
-                                 new neuron::Morphology( raw, transforms[i] )));
+            // store circuit + GID for transformed morphology
+            hash += _impl->getCircuitSource().getPath() +
+                       boost::lexical_cast< std::string >(i+1);
```

Comment:
You have to iterate over the GIDs at the same time you loop because i is just an index.